### PR TITLE
[WIP] REPL: TS support

### PR DESCRIPTION
**Pull request by bartlomieju**
Ref #1158

- can execute TS in REPL
- uses public `Deno.compile()` API to compile TS to JS - it spawns new compiler on each loop - it's not instant, but it's not that bad <1s (keeping compiler alive will definitely help, got my eyes on it)
- [TODO] TS diagnostic that are terminal errors are not pretty printed 
- [TODO] code is a mess 
- [TODO] copy-pasted file from std

### CODE REVIEW

File: cli/js/diff.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
+// TODO(bartlomieju): this file is copied from std/
```
Comment (ry) [inline cli/js/diff.ts]: What is the purpose of this "diff" module?

Comment (bartlomieju) [reply cli/js/diff.ts]: I implemented it just like ts-node does; compile TS to JS, diff changes from previous loop pass, execute newly added lines.

**Comment by realharry** [comment]
@ry Just to quickly follow up, and to add my perspective, I like typescript. It's not only because it supports type information. It provides other benefits as well (such as what babel does, essentially, for example). Javascript is a chaos. Figuring out what language features are supported in what runtimes (mostly in different browsers) is a nightmare, and it is probably unique to Javascript (that's why we need tools like babel, etc.). No other main programming languages have this issue (is this Java 8 or Java 9, does this code include C# 8 features? etc. are, I think, order of magnitude simpler problems than what we deal with in Javascript). Even in Node.js, do you know what ES6 or ES7/8/... features are supported in the current version of Node.js you are using? You literally have to google every time you need this information.

One of the advantages of using Typescript is it has a clear semantic versioning, on the language level. 

I think it is, can be, also a big benefits of using Deno. If I do 

```
$ deno --version
deno 0.31.0
v8 8.1.108
typescript 3.7.2
```

it tells me what version of typescript I'm using (more like, what version of deno, which supports what version of typescript).

Personally, I've been looking for a good repl "all my life". :) Not just a simple repl where I can just do '1+2', but more as a programming/prototyping environment. Deno REPL (if it supports Typescript) can be the one. (There are some programming languages that have good REPLs but, unfortunately, not for the ones I primarily use, C#, Go, Rust, Typescript, ...)

Can we maybe add an optional flag to enable the Typescript support in deno repl if the user chooses to do so?

**Comment by ry** [comment]
Optional flag sounds good to me.

**Comment by ry** [comment]
Just a bit more context for future researchers: There are many things we can do to improve the REPL without introducing the TS compiler. We should do those things first (e.g. improve inspect, tab completion). Once we're on par with Node's REPL, we can start looking into how to go beyond that by using typescript.

**Pull request closed without merging**